lnwallet/chainfee/filtermanager.go

```diff
@@ -0,0 +1,300 @@
+	if err != nil {
+		return nil, err
+	}
+
+	var peerResps []bitcoindPeerInfoResp
+	err = json.Unmarshal(resp, &peerResps)
+	if err != nil {
+		return nil, err
+	}
+
+	// We filter for outbound peers since it is harder for an attacker to
+	// be our outbound peer and therefore harder for them to manipulate us
+	// into broadcasting high-fee or low-fee transactions.
+	var outboundPeerFilters []SatPerKWeight
+	for _, peerResp := range peerResps {
+		if peerResp.Inbound {
+			continue
+		}
+
+		// The value that bitcoind returns for the "minfeefilter" field
+		// is in fractions of a bitcoin that represents the satoshis
+		// per KvB. We need to convert this fraction to whole satoshis
+		// by multiplying with COIN. Then we need to convert the
+		// sats/KvB to sats/KW.
+		//
+		// Convert the sats/KvB from fractions of a bitcoin to whole
+		// satoshis.
+		filterKVByte := SatPerKVByte(
+			peerResp.MinFeeFilter * btcutil.SatoshiPerBitcoin,
+		)
+
+		if !checkFilterBounds(filterKVByte) {
+			continue
+		}
+
+		// Convert KvB to KW and add it to outboundPeerFilters.
+		outboundPeerFilters = append(
+			outboundPeerFilters, filterKVByte.FeePerKWeight(),
+		)
+	}
+
+	// Check that we have enough data to use. We don't return an error as
+	// that would stop the querying goroutine.
+	if len(outboundPeerFilters) < minNumFilters {
+		return nil, nil
+	}
+
+	return outboundPeerFilters, nil
+}
+
+func fetchBtcdFilters(client *rpcclient.Client) ([]SatPerKWeight, error) {
+	resp, err := client.GetPeerInfo()
+	if err != nil {
+		return nil, err
+	}
+
+	var outboundPeerFilters []SatPerKWeight
+	for _, peerResp := range resp {
+		// We filter for outbound peers since it is harder for an
+		// attacker to be our outbound peer and therefore harder for
+		// them to manipulate us into broadcasting high-fee or low-fee
+		// transactions.
+		if peerResp.Inbound {
+			continue
+		}
+
+		// The feefilter is already in units of sat/KvB.
+		filter := SatPerKVByte(peerResp.FeeFilter)
+
+		if !checkFilterBounds(filter) {
+			continue
+		}
+
+		outboundPeerFilters = append(
+			outboundPeerFilters, filter.FeePerKWeight(),
+		)
+	}
+
+	// Check that we have enough data to use. We don't return an error as
+	// that would stop the querying goroutine.
+	if len(outboundPeerFilters) < minNumFilters {
+		return nil, nil
+	}
+
+	return outboundPeerFilters, nil
+}
+
+// updateWindow takes a slice of feefilter values and adds them to the moving
+// window.
+func (f *filterManager) updateWindow(feeFilters []SatPerKWeight) {
+	// If there are no elements, don't update.
+	numElements := len(feeFilters)
+	if numElements == 0 {
+		return
+	}
+
+	f.valuesMtx.Lock()
+	defer f.valuesMtx.Unlock()
+
+	if len(f.values)+numElements <= medianWindow {
+		// If there's room, we simply append.
+		f.values = append(f.values, feeFilters...)
+	} else {
+		// Otherwise, we'll roll off the excess elements.
+		numExcess := len(f.values) + numElements - medianWindow
+		f.values = f.values[numExcess:]
```

Comment:
I think this may memleak because the underlying buffer will grow unbounded just updating the slice metadata. It's possible that the append operation only does a partial copy when you overrun capacity space which would allow for reclamation. Someone wit better go-fu should check me on this.

Reply:
That's a good point. It doesn't leak very much because eventually it will reallocate a new underlying array (at around capacity 80) and GC the old one. It is a bit hacky though and the elements should be zero'd